File: mobile/src/main/java/com/novoda/materialised/hackernews/asynclistview/ModelledViewInflater.java
package com.novoda.materialised.hackernews.asynclistview;

import android.content.Context;
import android.view.View;
import android.view.ViewGroup;

import java.lang.reflect.Constructor;

final class ModelledViewInflater<V extends View & ModelledView> {
    private final Class<V> viewClass;

    ModelledViewInflater(Class<V> viewClass) {
        this.viewClass = viewClass;
    }

    V inflateUsing(ViewGroup parent) {
        Constructor<V> constructor;
        try {
            constructor = viewClass.getDeclaredConstructor(Context.class);
            return constructor.newInstance(parent.getContext());
        } catch (Exception e) { // this is horrendous, but multi-catching gives a compile error on API < 19
            e.printStackTrace();
            throw new RuntimeException("Couldn't inflate ModelledView instance!");
        }
    }
}

File: mobile/src/main/java/com/novoda/materialised/hackernews/asynclistview/ModelledViewInflater.kt
package com.novoda.materialised.hackernews.asynclistview

import android.content.Context
import android.view.View
import android.view.ViewGroup

import java.lang.reflect.Constructor

internal class ModelledViewInflater<V>(private val viewClass: Class<V>) where V : View, V : ModelledView<*> {

    fun inflateUsing(parent: ViewGroup): V {
        val constructor: Constructor<V>
        try {
            constructor = viewClass.getDeclaredConstructor(Context::class.java)
            return constructor.newInstance(parent.context)
        } catch (e: Exception) { // this is horrendous, but multi-catching gives a compile error on API < 19
            e.printStackTrace()
            throw RuntimeException("Couldn't inflate ModelledView instance!")
        }
    }
}
